src/client/executor.rs

```diff
@@ -89,8 +86,50 @@ impl Client {
             }
         };
 
+        let first_error = match self
+            .execute_operation_on_connection(&op, &mut conn, &mut session)
+            .await
+        {
+            Ok(result) => {
+                return Ok(result);
+            }
+            Err(err) => {
+                self.inner
+                    .topology
+                    .handle_post_handshake_error(err.clone(), conn, server)
+                    .await;
+                // TODO RUST-90: Do not retry if session is in a transaction
+                if self.inner.options.retry_reads == Some(false)
+                    || !op.is_read_retryable()
+                    || !err.is_read_retryable()
+                {
+                    return Err(err);
+                } else {
+                    err
+                }
+            }
+        };
+
+        let server = match self.select_server(op.selection_criteria()).await {
```

Comment:
I'm not sure if it's any clearer so feel free to not change this, but you can use `map_err` here in conjunction with the `?` operator as an alternative:
```rust
let server = self.select_server(op.selection_criteria()).await.map_err(|_| first_error)?;
```

Reply:
I tried this out, but it doesn't compile unless `first_error` is replaced with `first_error.clone()` in the closure to avoid the closure taking ownership of the error. I think I'll leave as-is to avoid an unnecessary clone.